Select primary font for missing glyph for PUA characters

vitorroriz opened the pull request:
#### df0f17533ab4c17991214087049e647a68754ce5
<pre>
Select primary font for missing glyph for PUA characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=273233">https://bugs.webkit.org/show_bug.cgi?id=273233</a>
<a href="https://rdar.apple.com/124036223">rdar://124036223</a>

Reviewed by Antti Koivisto.

We used to render the missing glyph for PUA characters (see [1])
with the lastResortFontFallback, instead of the primary font.
This is done by FontCascadeFonts::glyphDataForVariant(...)
producing a GlyphData object with such font for the missing glyph (glyph 0).

The ComplexTextController, which handles complex text, would always
use the primary font for getting metrics for the missing glyph. When rendering
PUA characters we would use the simple path (WidthIterator is used instead
of ComplexTextController), selecting the last resort font both for layout and painting.

However, when computing the caret position, we still use the ComplexTextController
because WidthIterator can&apos;t handle partial runs with ligatures (for example).

This would result in WidthIterator being used for layout/painting and selecting the
lastResortFont, and ComplexTextController being used for calculating the caret position,
selecting the primary font.

We are fixing it by updating the special case for PUA character and selecting
the primary font for such case, instead of the lastResortFont. This also has
the benefit of enforcing interoperability with other browser, as Blink
and Gecko seem to also use the primary font selected by the author.

[1] <a href="https://commits.webkit.org/269524@main">https://commits.webkit.org/269524@main</a>

* LayoutTests/platform/mac/fast/text/softbank-emoji-expected.txt:
* LayoutTests/platform/wpe/fast/text/softbank-emoji-expected.txt:
- Rebasing this test since they contain PUA characters and font will change.

* LayoutTests/TestExpectations:
- We should pass this test now.

* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForVariant):
- Using primary font instead of last resort font for glyph 0 when PUA characters can&apos;t be mapped by selected fonts.

Canonical link: <a href="https://commits.webkit.org/278221@main">https://commits.webkit.org/278221@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d420261bb3d173133d808c52455c8434e76ae6ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52860 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/294 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51923 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26523 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40496 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51718 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26478 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42733 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21611 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43927 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7989 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45828 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54437 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24704 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20876 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47871 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25972 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46899 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10943 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/graphics/FontCascadeFonts.h:
```diff
@@ -75,6 +75,7 @@ class FontCascadeFonts : public RefCounted<FontCascadeFonts> {
     const WidthCache& widthCache() const { return m_widthCache; }
 
     const Font& primaryFont(FontCascadeDescription&);
+    const Font& cachedPrimaryFont(const FontCascadeDescription&);
```
The naming here doesn't really communicate what is the difference between these very similar sounding functions. Both actually return a "cached" font.

vitorroriz replied:
How about `ensurePrimaryFont` and `primaryFont`?

I could call `primaryFont` something like `getPrimaryFontAndResolveReferences` but I'm leaning a bit against changing `primaryFont` name as the resolution of "font-size-adjust from-font".

Otherwise we could have 2 overloads for the function, one that receives a const object and one that receives a mutable object.

vitorroriz replied:
I ended up splitting this into another commit as it is not directly related to this bug: https://github.com/WebKit/WebKit/pull/27829

anttijk commented on Source/WebCore/platform/graphics/FontCascadeFonts.h:
```diff
@@ -128,6 +129,8 @@ class FontCascadeFonts : public RefCounted<FontCascadeFonts> {
     std::optional<Ref<Thread>> m_thread;
 #endif
+
+    bool isFontSizeAdjustFromFontResolved { false };
```
Class members should use m_ naming. 
Something like m_hasResolvedFontSizeAdjust might read better?

vitorroriz committed:
Select primary font for missing glyph for PUA characters
https://bugs.webkit.org/show_bug.cgi?id=273233
rdar://124036223

Reviewed by Antti Koivisto.

We used to render the missing glyph for PUA characters (see [1])
with the lastResortFontFallback, instead of the primary font.
This is done by FontCascadeFonts::glyphDataForVariant(...)
producing a GlyphData object with such font for the missing glyph (glyph 0).

The ComplexTextController, which handles complex text, would always
use the primary font for getting metrics for the missing glyph. When rendering
PUA characters we would use the simple path (WidthIterator is used instead
of ComplexTextController), selecting the last resort font both for layout and painting.

However, when computing the caret position, we still use the ComplexTextController
because WidthIterator can't handle partial runs with ligatures (for example).

This would result in WidthIterator being used for layout/painting and selecting the
lastResortFont, and ComplexTextController being used for calculating the caret position,
selecting the primary font.

We are fixing it by updating the special case for PUA character and selecting
the primary font for such case, instead of the lastResortFont. This also has
the benefit of enforcing interoperability with other browser, as Blink
and Gecko seem to also use the primary font selected by the author.

[1] https://commits.webkit.org/269524@main

* LayoutTests/platform/mac/fast/text/softbank-emoji-expected.txt:
* LayoutTests/platform/wpe/fast/text/softbank-emoji-expected.txt:
- Rebasing this test since they contain PUA characters and font will change.

* LayoutTests/TestExpectations:
- We should pass this test now.

* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForVariant):
- Using primary font instead of last resort font for glyph 0 when PUA characters can't be mapped by selected fonts.

Canonical link: https://commits.webkit.org/278221@main
